### ISSUE TITLE
`{include}`'d template ignores that `{include}` params may be different

### ISSUE THREAD

**Issue by spaze** [efabrica-team/phpstan-latte#391]
I have a template with
```
	{if $order == 'desc'}  // <-- line 40
		{var $dateLineCheck = $now && $training->getStart() < $now}  // <-- line 41
```
([source](https://github.com/spaze/michalspacek.cz/blob/82d0b7221e2b4251626acea8fbff3563991cbee0/site/app/Admin/Presenters/templates/Trainings/common/dateList.latte#L40-L41))

The template is included in some other templates.
For example like this:
```
{include "common/dateList.latte", order: 'desc', upcomingIds: $upcomingIds, now: $now}
```
([source](https://github.com/spaze/michalspacek.cz/blob/82d0b7221e2b4251626acea8fbff3563991cbee0/site/app/Admin/Presenters/templates/Trainings/default.latte#L9))
or like this:
```
    {include "common/dateList.latte", order: 'desc', upcomingIds: [], now: null}
```
([source](https://github.com/spaze/michalspacek.cz/blob/82d0b7221e2b4251626acea8fbff3563991cbee0/site/app/Admin/Presenters/templates/Trainings/pastWithPersonalData.latte#L6))
note the difference `now: null` vs. `now: $now` (where `$now` is created as `$this->template->now = new DateTime();`)

In the first case the error message says
```
 ------ ---------- -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   Compiled   Admin/Presenters/templates/Trainings/common/dateList.latte rendered from MichalSpacekCz\Admin\Presenters\TrainingsPresenter::default included in Admin/Presenters/templates/Trainings/default.latte
         line       See compiled template: /tmp/phpstan-latte/app/Admin/Presenters/templates/Trainings/common/dateList.latte.7844c3b4b8a153b2b6150427a6cbf075.php
 ------ ---------- -----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  40     74         If condition is always true.
  41     76         Left side of && is always true.
```
and in the second
```
 ------ ---------- -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  Line   Compiled   Admin/Presenters/templates/Trainings/common/dateList.latte rendered from MichalSpacekCz\Admin\Presenters\TrainingsPresenter::pastWithPersonalData included in Admin/Presenters/templates/Trainings/pastWithPersonalData.latte
         line       See compiled template: /tmp/phpstan-latte/app/Admin/Presenters/templates/Trainings/common/dateList.latte.57d7b0777d244e57e808707cb870ee8e.php
 ------ ---------- -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
  40     74         If condition is always true.
  41     76         Left side of && is always false.
```
(`order: 'desc'` on line 40 is changed in some other `{include}` line to `order: 'asc'`)

In both cases the template is compiled to:

```php
            /* line 40 */
            if ($order == 'desc') {
                /* line 41 */
                $dateLineCheck = $now && $training->getStart() < $now;
```
but in the first case the compiled source contains:
```
        /** @var DateTime $now */
        /** @var array<int, int> $upcomingIds */
        /** @var 'desc' $order */
```
and in the second case:
```php
        /** @var 'desc' $order */
        /** @var array{} $upcomingIds */
        /** @var null $now */
```
which is correct only in the isolated case and not when the template is reused and the parameter may be different.

I don't know what to do with that, if it's a missing feature or a bug, or if I should refactor my code.

**Comment by lulco** [comment]
We have to check this. Maybe @MartinMystikJonas will know more about it.

**Comment by MartinMystikJonas** [comment]
I have few similar cases in our code too.

It is one of limitations because we always analyse (included) template with given combination of passed parameters.

If we would merge parameters from different calls/includes it would cause even more weird false-positives. Because it would check invalid combinations. Like some bool flag deciding if render given part of template but with null values for data rendered in that part.

I solve it by ignoring these errors.

**Comment by spaze** [comment]
Thanks, I could probably change that and move the variables to be set in presenters so there are no `foo: null` in `{include}`s but it doesn't seem to solve the error because I think it's the same or similar as in #390.

**Comment by lulco** [comment]
Yes it is kind of similar issue. Here you should somehow use ?DateTime type for variable $now. But I understand it could be a lot of rewrite

**Comment by lulco** [comment]
Also I think some shared templates are large and contains more parts - each somehow conditionally displayed. So very often you have to use some flags (true/false) od null for nullable types. It would probably work neter (for phpstan latte) if you split those shared templates to multiple smaller templates and include only those you need.

**Comment by spaze** [comment]
> Also I think some shared templates are large and contains more parts

That's totally possible :-) Do you have an example which template you think it's way too huge?

**Comment by lulco** [comment]
I didn't work with your app for a while so I don’t remember. I just had that feeling

**Comment by spaze** [comment]
I think this can also be closed by #392, it's a similar issue as #390 anyway.
